_posts/2020-02-26-demo-fraud-detection-2.md

```diff
@@ -0,0 +1,192 @@
+
+```java
+public static final MapStateDescriptor<Integer, Rule> RULES_STATE_DESCRIPTOR =
+        new MapStateDescriptor<>(
+            "keys", BasicTypeInfo.INT_TYPE_INFO, TypeInformation.of(Rule.class));
+```
+
+Connecting to `rulesStream` causes some changes in the signature of the processing functions. Presented in a slightly simplified way in the previous article as `ProcessFunction`, `DynamicKeyFunction` is in reality `BroadcastProcessFunction`.
+
+```java
+public abstract class BroadcastProcessFunction<IN1, IN2, OUT> {
+
+    public abstract void processElement(IN1 value, ReadOnlyContext ctx, Collector<OUT> out) throws Exception;
+
+    public abstract void processBroadcastElement(IN2 value, Context ctx, Collector<OUT> out) throws Exception;
+
+}
+```
+
+The difference is the addition of a `processBroadcastElement` method through which the control messages will arrive. The version of `DynamicKeyFunciton` that allows  modifying the list of data distribution keys at runtime by sending rules definitions to `rules` Kafka topic (`RulesSource`) looks as follows:
+
+```java
+public class DynamicKeyFunction
+    extends BroadcastProcessFunction<Transaction, Rule, Keyed<Transaction, String, Integer>> {
+
+
+  @Override
+  public void processBroadcastElement(Rule rule,
+                                     Context ctx,
+                                     Collector<Keyed<Transaction, String, Integer>> out) {
+    BroadcastState<Integer, Rule> broadcastState = ctx.getBroadcastState(RULES_STATE_DESCRIPTOR);
+    broadcastState.put(rule.getRuleId(), rule);
+  }
+
+  @Override
+  public void processElement(Transaction event,
+                           ReadOnlyContext ctx,
+                           Collector<Keyed<Transaction, String, Integer>> out){
+    ReadOnlyBroadcastState<Integer, Rule> rulesState =
+                                  ctx.getBroadcastState(RULES_STATE_DESCRIPTOR);
+    for (Map.Entry<Integer, Rule> entry : rulesState.immutableEntries()) {
+        final Rule rule = entry.getValue();
+        out.collect(
+          new Keyed<>(
+            event, KeysExtractor.getKey(rule.getGroupingKeyNames(), event), rule.getRuleId()));
+    }
+  }
+}
+```
+
+In the above code `processElement()` receives Transactions and `processBroadcastElement()` receives Rules updates. When a new rule is created, it is distributed and saved in all parallel instances of the operator using `processBroadcastState`, as depicted in Figure 6. We use Rule's ID as the key to store and reference individual rules. Instead of iterating over a hardcoded `List<Rules>`, iteration is performed over entries in the dynamically updated broadcast state.
```

Comment:
```suggestion
In the code above, `processElement()` receives Transactions and `processBroadcastElement()` receives Rules updates. When a new rule is created, it is distributed and saved in all parallel instances of the operator using `processBroadcastState`, as depicted in Figure 6. We use a Rule's ID as the key to store and reference individual rules. Instead of iterating over a hardcoded `List<Rules>`, iteration is performed over entries in the dynamically updated broadcast state.
```
One thing that is still a bit misleading is that "Figure 6" does not show `processBroadcastState`...

Reply:
" `processBroadcastState`, as depicted in Figure 6." ->
" `processBroadcastState`, in such a manner as depicted in Figure 6." ?

Reply:
no, maybe "it is distributed as in Figure 6 and saved in..."
(I was a bit confused by the placement - Fig. 6 shows the distribution, not the saving part)

Reply:
https://forum.wordreference.com/threads/see-the-above-below-picture-or-see-the-picture-above-below.1301337/